src/eth/rpc/rpc_subscriptions.rs

```diff
@@ -300,6 +301,29 @@ pub struct RpcSubscriptionsConnected {
 }
 
 impl RpcSubscriptionsConnected {
+    /// Checks the number of subscriptions for a given client.
+    pub async fn check_client_subscriptions(&self, max_subscriptions: u32, client: &RpcClientApp) -> Result<(), RpcError> {
+        let pending_txs = self.pending_txs.read().await.values().filter(|s| s.client == *client).count();
+        let new_heads = self.new_heads.read().await.values().filter(|s| s.client == *client).count();
+        let logs = self
+            .logs
+            .read()
+            .await
+            .values()
+            .flat_map(HashMap::values)
+            .filter(|s| s.client == *client)
+            .count();
+        tracing::info!(%pending_txs, %new_heads, %logs, "current client subscriptions");
+
+        if pending_txs + new_heads + logs >= max_subscriptions as usize {
+            return Err(RpcError::SubscriptionInvalid {
```

Comment:
You need to create a new entry in the error enum for this validation.
The error being used here will give the wrong reason for the client.